File: alice/gamma/sdk/python/gamma_sdk/sdk/card.py
# coding:utf-8
import attr
import json

import alice.gamma.sdk.api.card_pb2 as api


@attr.s(frozen=True)
class Header:
    text = attr.ib(type=str)

    def to_proto(self):
        return api.CardHeader(
            text=self.text,
        )


@attr.s(frozen=True)
class Button:
    text = attr.ib(type=str)
    url = attr.ib(type=str)
    payload = attr.ib(default=None)

    def to_proto(self):
        return api.CardButton(
            text=self.text,
            url=self.url,
            payload=json.dumps(self.payload).encode() if self.payload else None,
        )


@attr.s(frozen=True)
class Item:
    image_id = attr.ib(type=str)
    title = attr.ib(type=str)
    description = attr.ib(type=str)
    button = attr.ib(type=Button, default=None)

    def to_proto(self):
        return api.CardItem(
            imageId=self.image_id,
            title=self.title,
            description=self.description,
            button=self.button.to_proto() if self.button else None,
        )

    @staticmethod
    def items_to_proto(items):
        return [item.to_proto() for item in items]


@attr.s(frozen=True)
class Footer:
    text = attr.ib(type=str)
    button = attr.ib(type=Button, default=None)

    def to_proto(self):
        return api.CardFooter(
            text=self.text,
            button=self.button.to_proto() if self.button else None,
        )


@attr.s(frozen=True)
class Card:
    type = attr.ib(type=str)
    image_id = attr.ib(type=str)
    title = attr.ib(type=object)
    description = attr.ib(type=str)
    button = attr.ib(type=Button, default=None)
    header = attr.ib(type=Header, default=None)
    footer = attr.ib(type=Footer, default=None)
    items = attr.ib(type=list, default=list())

    def add_items(self, *items):
        self.items.extend(items)

    def to_proto(self):
        return api.Card(
            type=self.type,
            imageId=self.image_id,
            title=self.title,
            description=self.description,
            button=self.button.to_proto() if self.button else None,
            header=self.header.to_proto() if self.header else None,
            footer=self.footer.to_proto() if self.footer else None,
            items=Item.items_to_proto(self.items) if self.items else None,
        )

File: alice/gamma/sdk/golang/card.go
package sdk

import (
	"bytes"
	"encoding/json"
	"image"
	"image/png"
	"net/http"

	"a.yandex-team.ru/alice/gamma/sdk/api"
)

type CardHeader struct {
	Text string
}

func (header *CardHeader) toProto() *api.CardHeader {
	return &api.CardHeader{Text: header.Text}
}

type CardButton struct {
	Text    string
	Payload interface{}
	URL     string
}

func (button *CardButton) toProto() (apiButton *api.CardButton, err error) {
	apiButton = &api.CardButton{
		Text: button.Text,
		Url:  button.URL,
	}
	if button.Payload != nil {
		apiButton.Payload, err = json.Marshal(button.Payload)
		if err != nil {
			return nil, err
		}
	}
	return apiButton, nil
}

type CardItem struct {
	ImageID     string
	Title       string
	Description string
	Button      *CardButton
}

func (item *CardItem) toProto() (protoItem *api.CardItem, err error) {
	protoItem = &api.CardItem{
		ImageId:     item.ImageID,
		Title:       item.Title,
		Description: item.Description,
	}
	if item.Button != nil {
		protoItem.Button, err = item.Button.toProto()
		if err != nil {
			return nil, err
		}
	}
	return protoItem, nil
}

func cardItemsToProto(cardItems []CardItem) (protoItems []*api.CardItem, err error) {
	protoItems = make([]*api.CardItem, len(cardItems))
	for i, item := range cardItems {
		protoItems[i], err = item.toProto()
		if err != nil {
			return nil, err
		}
	}
	return protoItems, nil
}

type CardFooter struct {
	Text   string
	Button *CardButton
}

func (footer *CardFooter) toProto() (protoFooter *api.CardFooter, err error) {
	protoFooter = &api.CardFooter{Text: footer.Text}
	if footer.Button != nil {
		protoFooter.Button, err = footer.Button.toProto()
		if err != nil {
			return nil, err
		}
	}
	return protoFooter, nil
}

type Card struct {
	Type        string
	ImageID     string
	Title       string
	Description string
	Button      *CardButton
	Header      *CardHeader
	Items       []CardItem
	Footer      *CardFooter
}

func (card *Card) AddCardItems(items ...CardItem) {
	card.Items = append(card.Items, items...)
}

func (card *Card) toProto() (protoCard *api.Card, err error) {
	protoCard = &api.Card{
		Type:        card.Type,
		ImageId:     card.ImageID,
		Title:       card.Title,
		Description: card.Description,
	}
	if card.Header != nil {
		protoCard.Header = card.Header.toProto()
	}
	if card.Button != nil {
		protoCard.Button, err = card.Button.toProto()
		if err != nil {
			return nil, err
		}
	}
	if card.Items != nil {
		protoCard.Items, err = cardItemsToProto(card.Items)
		if err != nil {
			return nil, err
		}
	}
	if card.Footer != nil {
		protoCard.Footer, err = card.Footer.toProto()
		if err != nil {
			return nil, err
		}
	}
	return protoCard, nil
}

func PostImage(image image.Image, url string) (string, error) {
	buffer := new(bytes.Buffer)
	var err error

	if err = png.Encode(buffer, image); err != nil {
		return "", err
	}

	response, err := http.Post(url, "image/png", buffer)
	if err != nil {
		return "", err
	}

	var responseBody map[string]interface{}
	if err = json.NewDecoder(response.Body).Decode(&responseBody); err != nil {
		return "", err
	}

	return responseBody["url"].(string), nil
}
